src/Features/CSharp/Portable/AddBraces/CSharpAddBracesDiagnosticAnalyzer.cs

```diff
@@ -86,5 +108,172 @@ public void AnalyzeNode(SyntaxNodeAnalysisContext context)
                 properties: null,
                 SyntaxFacts.GetText(firstToken.Kind())));
         }
+
+        /// <summary>
+        /// <para>In general, statements are considered multiline if any of the following span more than one line:</para>
+        /// <list type="bullet">
+        /// <item><description>The part of the statement preceding the embedded statement</description></item>
+        /// <item><description>The embedded statement itself</description></item>
+        /// <item><description>The part of the statement following the embedded statement</description></item>
+        /// </list>
+        /// <para>The third condition is not checked for <c>else</c> clauses because they are only considered multiline
+        /// when their embedded statement is multiline.</para>
+        /// </summary>
+        private static bool IsConsideredMultiLine(SyntaxNode statement, SyntaxNode embeddedStatement)
+        {
+            // Early return if the entire statement fits on one line
+            if (FormattingRangeHelper.AreTwoTokensOnSameLine(statement.GetFirstToken(), statement.GetLastToken()))
+            {
+                // The entire statement fits on one line. Examples:
+                //
+                //   if (something) return;
+                //
+                //   while (true) something();
+                return false;
+            }
+
+            // Check the part of the statement preceding the embedded statement
+            if (!FormattingRangeHelper.AreTwoTokensOnSameLine(embeddedStatement.GetFirstToken(), embeddedStatement.GetLastToken()))
+            {
+                // The embedded statement does not fit on one line. Examples:
+                //
+                //   if (something)
+                //     obj.Method(   // <-- This embedded statement spans two lines.
+                //       arg);
+                return true;
+            }
+
+            // Check the embedded statement itself
+            var lastTokenBeforeEmbeddedStatement = embeddedStatement.GetFirstToken().GetPreviousToken();
+            if (!FormattingRangeHelper.AreTwoTokensOnSameLine(statement.GetFirstToken(), lastTokenBeforeEmbeddedStatement))
+            {
+                // The part of the statement preceding the embedded statement does not fit on one line. Examples:
+                //
+                //   for (int i = 0; // <-- The initializer/condition/increment are on separate lines
+                //        i < 10;
+                //        i++)
+                //     SomeMethod();
+                return true;
+            }
+
+            // Check the part of the statement following the embedded statement, but only if it exists
```

Comment:
maybe add "and only if it isn't an else-clause (as per the specification defined above)."

Reply:
➡️ Now updated